core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java

```diff
@@ -19,115 +19,299 @@
+        return Objects.hash(center, geoDistance, optimizeBbox, distance, coerce, ignoreMalformed);
+    }
+
+    @Override
+    public boolean doEquals(GeoDistanceQueryBuilder other) {
+        return Objects.equals(fieldName, other.fieldName) &&
+                (distance == other.distance) &&
+                (coerce == other.coerce) &&
+                (ignoreMalformed == other.ignoreMalformed) &&
+                Objects.equals(center, other.center) &&
+                Objects.equals(optimizeBbox, other.optimizeBbox) &&
+                Objects.equals(geoDistance, other.geoDistance);
+    }
+
+    @Override
+    protected GeoDistanceQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        String fieldName = in.readString();
+        GeoDistanceQueryBuilder result = new GeoDistanceQueryBuilder(fieldName);
+        result.distance = in.readDouble();
+        result.coerce = in.readBoolean();
+        result.ignoreMalformed = in.readBoolean();
+        result.center = GeoPoint.readGeoPointFrom(in);
+        result.optimizeBbox = in.readString();
+        result.geoDistance = GeoDistance.readGeoDistanceFrom(in);
+        return result;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeDouble(distance);
+        out.writeBoolean(coerce);
+        out.writeBoolean(ignoreMalformed);
+        center.writeTo(out);
+        out.writeString(optimizeBbox);
+        geoDistance.writeTo(out);
+    }
+
+    @Override
+        public QueryValidationException validate() {
+            QueryValidationException validationException = null;
+            // validation was not available prior to 2.x, so to support bwc percolation queries we only ignore_malformed on 2.x created indexes
+            if (!ignoreMalformed) {
+                if (Numbers.isValidDouble(center.getLat()) == false) {
+                    validationException = addValidationError("center point latitude is invalid number: " + center.getLat(), validationException);
+                }
+                if (Numbers.isValidDouble(center.getLon()) == false) {
+                    validationException = addValidationError("center point longitude is invalid number: " + center.getLon(), validationException);
+                }
+            }
+
+            if (Strings.isEmpty(fieldName)) {
+                validationException = addValidationError("field name must be non-null and non-empty", validationException);
+            }
+            if (optimizeBbox != null && !(optimizeBbox.equals("none") || optimizeBbox.equals("memory") || optimizeBbox.equals("indexed"))) {
+                validationException = QueryValidationException.addValidationError(NAME, "optimizeBbox must be one of [none, memory, indexed]", validationException);
+            }
+            return validationException;
+        }
+
+    QueryValidationException validate(boolean indexCreatedBeforeV2_0) {
```

Comment:
@colings86 suggested to use a different method name over in #11969 which I think makes sense.